proposals/0000-astro-url-helper.md

```diff
@@ -0,0 +1,118 @@
+- Start Date: 2022-07-18
+- Reference Issues: N/A
+- Implementation PR: https://github.com/withastro/astro/pull/3959
+
+# Summary
+
+Create a general-purpose `Astro.url` helper. Deprecate `Astro.canonicalURL` to use `Astro.url` instead.
+
+# Example
+
+```js
+// Before:
+const currentPathname = Astro.canonicalURL.pathname;
+const currentPathnameAlternativeAPI = new URL(Astro.request.url).pathname;
+const origin = new URL(Astro.request.url).origin;
+// After:
+const currentPathname = Astro.url.pathname;
+const origin = Astro.url.origin;
+```
+
+```js
+// Before:
+const canonicalURL = Astro.canonicalURL;
+// After:
+const canonicalURL = new URL(Astro.url.pathname, Astro.site);
+```
+
+```js
+// More complex example (from Astro monorepo)
+// Before:
+const canonicalURL = new URL(new URL(Astro.request.url).pathname, Astro.site ?? `http://example.com`);
+// After:
+const canonicalURL = new URL(Astro.url.pathname, Astro.site ?? `http://example.com`);
+```
+
+# Motivation
+
+![Example diagram showing complexity of current solutions]()
+
+## Renaming/Replacing `Astro.canonicalURL`
+
+It's difficult for a framework to understand the concept of "canonical URL", since canonical is a concept that can mean different things to different projects. For example:
+
+- How does Astro know if multiple pages should have a single canonical URL?
+- In SSR mode, is every request caught by `[...foo].astro` canonically unique?
+
+In addition, there was confusion over the canonical domain used:
+
+- If `Astro.site` is set, what domain does `Astro.canonicalURL` use?
+- If `Astro.site` is not set, what domain does `Astro.canonicalURL` use?
+
+The result is that we have introduced the possibility that in some cases Astro is "lying" to the user about a URL being canonical when it may not be. This is most common when in SSR (where we don't know the built URLs ahead of time) and when Astro.site is not set (where we use the current origin instead of known production origin).
```

Comment:
I think the SSR point needs a bit more clarification for what you mean. I *think* what you are saying is that you, as a developer can "vouch" for the correctness of a canonical URL in SSG mode because you can see the result before it's merged into production. But I had to think about the sentence a lot to guess at the meaning (if that's even correct)

Reply:
Good catch, I rewrote this section.